include/class_loader/class_loader.hpp

```diff
@@ -122,12 +122,21 @@ class ClassLoader
    * @return A std::shared_ptr<Base> to newly created plugin object
    */
   template<class Base>
-  std::shared_ptr<Base> createInstance(const std::string & derived_class_name)
+  std::shared_ptr<Base> createInstance(
+    const std::string & derived_class_name,
+    bool is_shared_ptr = false)
   {
-    return std::shared_ptr<Base>(
-      createRawInstance<Base>(derived_class_name, true),
-      std::bind(&ClassLoader::onPluginDeletion<Base>, this, std::placeholders::_1)
-    );
+    if (is_shared_ptr) {
+      return std::shared_ptr<Base>(
+        createRawInstance<Base>(derived_class_name, true),
+        std::bind(&ClassLoader::onPluginDeletion<Base>, shared_from_this(), std::placeholders::_1)
+      );
+    } else {
+      return std::shared_ptr<Base>(
+        createRawInstance<Base>(derived_class_name, true),
+        std::bind(&ClassLoader::onPluginDeletion<Base>, this, std::placeholders::_1)
+      );
+    }
```

Comment:
@ahcorde nit^N, safe to ignore: 
```suggestion
    return std::shared_ptr<Base>(
      createRawInstance<Base>(derived_class_name, true),
      std::bind(&ClassLoader::onPluginDeletion<Base>, 
                is_shared_ptr ? shared_from_this() : this, 
                std::placeholders::_1));
```